In [1]:
import numpy as np
import pandas as pd

In [2]:
data = pd.read_csv("/Users/jeewonkim/Desktop/creditcard.csv")

In [3]:
data.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [4]:
data.shape

(284807, 31)

In [5]:
X = data.drop(['Class'], axis=1)
y = data['Class']

In [6]:
X.shape

(284807, 30)

In [7]:
y.shape

(284807,)

### Stratified train test split

In [8]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X,y,stratify = y,random_state = 42)

In [9]:
print(X_train.shape)
print(X_test.shape)

(213605, 30)
(71202, 30)


### (1) Create a classification model (supervised learning) (20pts)

In [10]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

tree = DecisionTreeClassifier(max_depth = 20).fit(X_train,y_train)
y_test_hat = tree.predict(X_test)

In [11]:
accuracy_score(y_test,y_test_hat)

0.9991713715906857

### (2) Apply two different methods to overcome the class imbalance problem (use imblearn package https://imbalanced-learn.org/ ) (20pts)


In [12]:
pip install sklearn

Note: you may need to restart the kernel to use updated packages.


In [13]:
pip install imblearn

Note: you may need to restart the kernel to use updated packages.


In [14]:
pip install _ClassNamePrefixFeaturesOutMixin

ERROR: Invalid requirement: '_ClassNamePrefixFeaturesOutMixin'
Note: you may need to restart the kernel to use updated packages.


In [15]:
import imblearn

In [16]:
# Random Undersampling

from imblearn.under_sampling import RandomUnderSampler
X_resampled_under, y_resampled_under = RandomUnderSampler(random_state = 0).fit_resample(X,y)
y_resampled_under.value_counts()

1    492
0    492
Name: Class, dtype: int64

In [17]:
X_train, X_test, y_train, y_test = train_test_split(X_resampled_under,y_resampled_under,stratify = y_resampled_under,random_state = 42)

In [18]:
tree = DecisionTreeClassifier(max_depth = 20).fit(X_train,y_train)
y_test_hat = tree.predict(X_test)

In [19]:
accuracy_score(y_test,y_test_hat)

0.9105691056910569

In [20]:
# Random Oversampling

from imblearn.over_sampling import RandomOverSampler
X_resampled_over, y_resampled_over = RandomOverSampler(random_state = 0).fit_resample(X,y)
y_resampled_over.value_counts()

1    284315
0    284315
Name: Class, dtype: int64

In [21]:
X_train, X_test, y_train, y_test = train_test_split(X_resampled_over,y_resampled_over,stratify = y_resampled_over,random_state = 42)

In [22]:
tree = DecisionTreeClassifier(max_depth = 20).fit(X_train,y_train)
y_test_hat = tree.predict(X_test)

In [23]:
accuracy_score(y_test,y_test_hat)

0.9991699376749814

### (3) Create an unsupervised anomaly detection model (isolation forest) (20pts)

In [42]:
data_isolated = data.drop(data[data.Class == 1].index)
data_outliers = data.drop(data[data.Class == 0].index)

In [43]:
data_isolated

Time         V1         V2        V3        V4        V5  \
0            0.0  -1.359807  -0.072781  2.536347  1.378155 -0.338321   
1            0.0   1.191857   0.266151  0.166480  0.448154  0.060018   
2            1.0  -1.358354  -1.340163  1.773209  0.379780 -0.503198   
3            1.0  -0.966272  -0.185226  1.792993 -0.863291 -0.010309   
4            2.0  -1.158233   0.877737  1.548718  0.403034 -0.407193   
...          ...        ...        ...       ...       ...       ...   
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
0       0.462388  0.239599  0.098698  0.363787  ... -0.018307  0.277838   
1      -0.082361 -0.078803  0.085102 -0.255425  ... -0.225775 -0.638672   
2       1.800499  0.791461  0.247676 -1.514654  ...  0.247998  0.771679   
3       1.247203  0.237609  0.377436 -1.387024  ... -0.108300  0.005274   
4       0.095921  0.592941 -0.270533  0.817739  ... -0.009431  0.798278   
...          ...       ...       ...       ...  ...       ...       ...   
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
0      -0.110474  0.066928  0.128539 -0.189115  0.133558 -0.021053  149.62   
1       0.101288 -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69   
2       0.909412 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752  378.66   
3      -0.190321 -1.175575  0.647376 -0.221929  0.062723  0.061458  123.50   
4      -0.137458  0.141267 -0.206010  0.502292  0.219422  0.215153   69.99   
...          ...       ...       ...       ...       ...       ...     ...   
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
0           0  
1           0  
2           0  
3           0  
4           0  
...       ...  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[284315 rows x 31 columns]

In [44]:
X_isolated = data_isolated.drop(['Class'], axis=1)
y_isolated = data_isolated['Class']

In [45]:
y_isolated.value_counts()

0    284315
Name: Class, dtype: int64

In [46]:
X_outliers = data_outliers.drop(['Class'], axis=1)
y_outliers = data_outliers['Class']

In [47]:
y_outliers.value_counts()

1    492
Name: Class, dtype: int64

In [48]:
X_train, X_test, y_train, y_test = train_test_split(X_isolated,y_isolated,stratify = y_isolated,random_state = 42)

In [49]:
from sklearn.ensemble import IsolationForest

clf = IsolationForest(max_samples = 100, contamination = 0.1, random_state =42)
clf.fit(X_train)
y_pred_train = clf.predict(X_train)
y_pred_test = clf.predict(X_test)
y_pred_outliers = clf.predict(X_outliers)

/Users/jeewonkim/opt/anaconda3/lib/python3.8/site-packages/sklearn/base.py:450: UserWarning: X does not have valid feature names, but IsolationForest was fitted with feature names
  warnings.warn(


In [56]:
print("Accuracy in test dataset : ", list(y_pred_test).count(1)/y_pred_test.shape[0])

Accuracy in test set :  0.8989997045540877


In [57]:
print("Accuracy in anomaly dataset:", list(y_pred_outliers).count(-1)/y_pred_outliers.shape[0])

Accuracy in anomaly dataset: 0.8699186991869918


### (4) Evaluate performance of the four models (20pts)

#### The results are above
- Accuracy in classification model : 0.9991713715906857
- Accuracy in undersampling : 0.9105691056910569
- Accuracy in oversampling : 0.9991699376749814
- Accuracy in isolation forest : 0.8989997045540877(in test dataset), 0.8699186991869918(in anomaly dataset)

### (5) Compare models

Among four models, the accuracy among four models was the highest in normal classification model.